TYPO3.Neos/Classes/TYPO3/Neos/Controller/Module/Management/WorkspacesController.php

```diff
@@ -491,6 +500,146 @@ protected function computeSiteChanges(Workspace $selectedWorkspace)
     }
 
     /**
+     * Retrieves the given node's corresponding node in the base workspace (that is, which would be overwritten if the
+     * given node would be published)
+     *
+     * @param NodeInterface $modifiedNode
+     * @return NodeInterface
+     */
+    protected function getOriginalNode(NodeInterface $modifiedNode)
+    {
+        $baseWorkspaceName = $modifiedNode->getWorkspace()->getBaseWorkspace()->getName();
+        $contextProperties = $modifiedNode->getContext()->getProperties();
+        $contextProperties['workspaceName'] = $baseWorkspaceName;
+        $contentContext = $this->contextFactory->create($contextProperties);
+
+        return $contentContext->getNodeByIdentifier($modifiedNode->getIdentifier());
+    }
+
+    /**
+     * Renders the difference between the original and the changed content of the given node and returns it, along
+     * with meta information, in an array.
+     *
+     * @param NodeInterface $changedNode
+     * @return array
+     */
+    protected function renderContentChanges(NodeInterface $changedNode)
+    {
+        $contentChanges = [];
+        $originalNode = $this->getOriginalNode($changedNode);
+        $changeNodePropertiesDefaults = $changedNode->getNodeType()->getDefaultValuesForProperties($changedNode);
+
+        $renderer = new HtmlArrayRenderer();
+        foreach ($changedNode->getProperties() as $propertyName => $changedPropertyValue) {
+            if ($originalNode === null && empty($changedPropertyValue) || (isset($changeNodePropertiesDefaults[$propertyName]) && $changedPropertyValue === $changeNodePropertiesDefaults[$propertyName])) {
+                continue;
+            }
+
+            $originalPropertyValue = ($originalNode === null ? null : $originalNode->getProperty($propertyName));
+
+            if ($changedPropertyValue === $originalPropertyValue && !$changedNode->isRemoved()) {
+                continue;
+            }
+
+            if (!is_object($originalPropertyValue) && !is_object($changedPropertyValue)) {
+                $originalSlimmedDownContent = $this->renderSlimmedDownContent($originalPropertyValue);
+                $changedSlimmedDownContent = $changedNode->isRemoved() ? '' : $this->renderSlimmedDownContent($changedPropertyValue);
+
+                $diff = new Diff(explode("\n", $originalSlimmedDownContent), explode("\n", $changedSlimmedDownContent), ['context' => 1]);
+                $diffArray = $diff->render($renderer);
+                $this->postProcessDiffArray($diffArray);
+
+                if (count($diffArray) > 0) {
+                    $contentChanges[$propertyName] = [
+                        'type' => 'text',
+                        'propertyLabel' => $this->getPropertyLabel($propertyName, $changedNode),
+                        'diff' => $diffArray
+                    ];
+                }
+            } elseif ($originalPropertyValue instanceof AssetInterface || $changedPropertyValue instanceof AssetInterface) {
```

Comment:
agreed; in general we really need to extract the diff functionality into a service once we start implementing more features.